### DISCUSSION TITLE
`Async::HTTP` vs `Net::HTTP`

### DISCUSSION THREAD

**Discussion by domrettig** [socketry/async-http#180] [General]
This seems like a silly question, but I haven't found an answer to it yet. I'm struggling to wrap my mind around the practical differences between `Async::HTTP` and other non-async HTTP clients. The docs show both the following scenarios as valid uses of `async`:
```ruby
Async do
  Net::HTTP.get(uri)
end
```
```ruby
Async do
  Async::HTTP::Internet.get(uri)
end
```
Both will perform the HTTP request asynchronously. So my question is what is the benefit of using `Async::HTTP::Internet` in this case? When should I use that over `Net::HTTP` or other HTTP libs?

**Comment by ioquatix** [comment]
- `Async::HTTP` supports persistent connections by default.
- `Async::HTTP` supports HTTP/2 (and eventually 3 when I get time to finish up the prototype).
- `Async::HTTP` has first class support for streaming request and response bodies.
- `Async::HTTP` supports trailers and interim responses.
- `Async::HTTP` supports both client and server side with the same abstractions.

Let me know if you have any more specific questions or comparisons you'd like help with.

**Reply by domrettig** [reply]
That's all very helpful -- thanks! So it seems like a more fully featured HTTP client/server than other options, but I think I'm still a bit confused on what benefits this library has when writing async code compared to those other options. It seems like I'd interact with it in generally the same way (like the example in my original post).

**Reply by trevorturk** [reply]
I think this is a good read on the subject: https://brunosutic.com/blog/async-ruby

Note that you don't even have to use the Async gem: https://brunosutic.com/blog/ruby-fiber-scheduler

/cc @bruno- thank you for these write-ups!
